File: app/src/main/java/com/example/task2/models/benchmarks/CollectionBenchmark.java
package com.example.task2.models.benchmarks;

import com.example.task2.R;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CopyOnWriteArrayList;

public class CollectionBenchmark implements Benchmark {
    private static final String SPECIFIC_NUMBER = "28";
    private final Random random = new Random();

    @Override
    public int getNumberOfColumns() {
        return 3;
    }

    @Override
    public List<CellOperation> createItemsList(boolean setRunning) {
        return Arrays.asList(
                new CellOperation(R.string.adding_in_the_beginning, R.string.arraylist, R.string.na, setRunning),
                new CellOperation(R.string.adding_in_the_beginning, R.string.linkedlist, R.string.na, setRunning),
                new CellOperation(R.string.adding_in_the_beginning, R.string.copyonwritearraylist, R.string.na, setRunning),
                new CellOperation(R.string.adding_in_the_middle, R.string.arraylist, R.string.na, setRunning),
                new CellOperation(R.string.adding_in_the_middle, R.string.linkedlist, R.string.na, setRunning),
                new CellOperation(R.string.adding_in_the_middle, R.string.copyonwritearraylist, R.string.na, setRunning),
                new CellOperation(R.string.adding_in_the_end, R.string.arraylist, R.string.na, setRunning),
                new CellOperation(R.string.adding_in_the_end, R.string.linkedlist, R.string.na, setRunning),
                new CellOperation(R.string.adding_in_the_end, R.string.copyonwritearraylist, R.string.na, setRunning),
                new CellOperation(R.string.search_by_value, R.string.arraylist, R.string.na, setRunning),
                new CellOperation(R.string.search_by_value, R.string.linkedlist, R.string.na, setRunning),
                new CellOperation(R.string.search_by_value, R.string.copyonwritearraylist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_beginning, R.string.arraylist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_beginning, R.string.linkedlist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_beginning, R.string.copyonwritearraylist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_middle, R.string.arraylist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_middle, R.string.linkedlist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_middle, R.string.copyonwritearraylist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_end, R.string.arraylist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_end, R.string.linkedlist, R.string.na, setRunning),
                new CellOperation(R.string.removing_in_the_end, R.string.copyonwritearraylist, R.string.na, setRunning)
        );
    }

    @Override
    public long measureTime(CellOperation item, int number) {
        List<String> list;
        if (item.type == R.string.arraylist) {
            list = new ArrayList<>(Collections.nCopies(number, "test"));
        } else if (item.type == R.string.linkedlist) {
            list = new LinkedList<>(Collections.nCopies(number, "test"));
        } else if (item.type == R.string.copyonwritearraylist) {
            list = new CopyOnWriteArrayList<>(Collections.nCopies(number, "test"));
        } else {
            throw new RuntimeException("Unsupported collection type");
        }

        long result;
        if (item.action == R.string.adding_in_the_beginning) {
            result = addStart(list);
        } else if (item.action == R.string.adding_in_the_middle) {
            result = addMiddle(list);
        } else if (item.action == R.string.adding_in_the_end) {
            result = addEnd(list);
        } else if (item.action == R.string.search_by_value) {
            result = searchByValue(list);
        } else if (item.action == R.string.removing_in_the_beginning) {
            result = removeStart(list);
        } else if (item.action == R.string.removing_in_the_middle) {
            result = removeMiddle(list);
        } else if (item.action == R.string.removing_in_the_end) {
            result = removeEnd(list);
        } else {
            throw new RuntimeException("Unsupported action type");
        }
        return result;
    }

    private long addStart(List<String> list) {
        long startTime, endTime;
        startTime = System.nanoTime();
        list.add(0, "test");
        endTime = System.nanoTime();
        return (endTime - startTime);
    }

    private long addMiddle(List<String> list) {
        long startTime, endTime;
        startTime = System.nanoTime();
        list.add(list.size() / 2, "test");
        endTime = System.nanoTime();
        return (endTime - startTime);
    }

    private long addEnd(List<String> list) {
        long startTime, endTime;
        startTime = System.nanoTime();
        list.add("test");
        endTime = System.nanoTime();
        return (endTime - startTime);
    }

    private long searchByValue(List<String> list) {
        int randomIndex = random.nextInt(list.size());
        list.set(randomIndex, SPECIFIC_NUMBER);
        long startTime, endTime;
        startTime = System.nanoTime();
        list.contains(SPECIFIC_NUMBER);
        endTime = System.nanoTime();
        return (endTime - startTime);
    }

    private long removeStart(List<String> list) {
        long startTime, endTime;
        startTime = System.nanoTime();
        list.remove(0);
        endTime = System.nanoTime();
        return (endTime - startTime);
    }

    private long removeMiddle(List<String> list) {
        long startTime, endTime;
        startTime = System.nanoTime();
        list.remove(list.size() / 2);
        endTime = System.nanoTime();
        return (endTime - startTime);
    }

    private long removeEnd(List<String> list) {
        long startTime, endTime;
        startTime = System.nanoTime();
        list.remove(list.size() - 1);
        endTime = System.nanoTime();
        return (endTime - startTime);
    }
}

File: app/src/main/java/com/example/task2/models/benchmarks/CollectionBenchmark.kt
package com.example.task2.models.benchmarks

import com.example.task2.BenchmarksApp
import com.example.task2.R
import com.example.task2.models.room.CollectionResult
import com.example.task2.models.room.CollectionResultDao
import java.util.Collections
import java.util.LinkedList
import java.util.Random
import java.util.concurrent.CopyOnWriteArrayList
import javax.inject.Inject

open class CollectionBenchmark : Benchmark {

    companion object {
        private const val SPECIFIC_NUMBER = "28"
    }

    override fun injectDependenciesRoom() {
        BenchmarksApp.instance?.roomComponent?.inject(this)
    }

    @Inject
    lateinit var collectionResultDao: CollectionResultDao

    private val random = Random()

    override fun setData(action: String, type: String, time: Long, input: Int) {
        val operationResult = CollectionResult(
                action = action,
                type = type,
                time = time,
                input = input
        )
        try {
            collectionResultDao.insertResult(operationResult)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun getData(): List<CollectionResult> {
        return collectionResultDao.getLastResults()
    }

    override fun getNumberOfColumns(): Int = 3

    override fun createItemsList(setRunning: Boolean): List<CellOperation> = listOf(
            CellOperation(R.string.adding_in_the_beginning, R.string.arraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.adding_in_the_beginning, R.string.linkedlist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.adding_in_the_beginning, R.string.copyonwritearraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.adding_in_the_middle, R.string.arraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.adding_in_the_middle, R.string.linkedlist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.adding_in_the_middle, R.string.copyonwritearraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.adding_in_the_end, R.string.arraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.adding_in_the_end, R.string.linkedlist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.adding_in_the_end, R.string.copyonwritearraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.search_by_value, R.string.arraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.search_by_value, R.string.linkedlist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.search_by_value, R.string.copyonwritearraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_beginning, R.string.arraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_beginning, R.string.linkedlist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_beginning, R.string.copyonwritearraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_middle, R.string.arraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_middle, R.string.linkedlist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_middle, R.string.copyonwritearraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_end, R.string.arraylist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_end, R.string.linkedlist, R.string.na.toLong(), setRunning),
            CellOperation(R.string.removing_in_the_end, R.string.copyonwritearraylist, R.string.na.toLong(), setRunning)
    )

    override fun measureTime(item: CellOperation, number: Int): Long {

        val list = when (item.type) {
            R.string.arraylist -> ArrayList(Collections.nCopies(number, "test"))
            R.string.linkedlist -> LinkedList(Collections.nCopies(number, "test"))
            R.string.copyonwritearraylist -> CopyOnWriteArrayList(Collections.nCopies(number, "test"))
            else -> throw RuntimeException("Unsupported collection type")
        }

        return when (item.action) {
            R.string.adding_in_the_beginning -> addStart(list)
            R.string.adding_in_the_middle -> addMiddle(list)
            R.string.adding_in_the_end -> addEnd(list)
            R.string.search_by_value -> searchByValue(list)
            R.string.removing_in_the_beginning -> removeStart(list)
            R.string.removing_in_the_middle -> removeMiddle(list)
            R.string.removing_in_the_end -> removeEnd(list)
            else -> throw RuntimeException("Unsupported action type")
        }
    }

    private fun addStart(list: MutableList<String>): Long {
        val startTime: Long = System.nanoTime()
        list.add(0, "test")
        return System.nanoTime() - startTime
    }

    private fun addMiddle(list: MutableList<String>): Long {
        val startTime: Long = System.nanoTime()
        list.add(list.size / 2, "test")
        return System.nanoTime() - startTime
    }

    private fun addEnd(list: MutableList<String>): Long {
        val startTime: Long = System.nanoTime()
        list.add("test")
        return System.nanoTime() - startTime
    }

    private fun searchByValue(list: MutableList<String>): Long {
        val randomIndex = random.nextInt(list.size)
        list[randomIndex] = SPECIFIC_NUMBER
        val startTime: Long = System.nanoTime()
        list.contains(SPECIFIC_NUMBER)
        return System.nanoTime() - startTime
    }

    private fun removeStart(list: MutableList<String>): Long {
        val startTime: Long = System.nanoTime()
        list.removeAt(0)
        return System.nanoTime() - startTime
    }

    private fun removeMiddle(list: MutableList<String>): Long {
        val startTime: Long = System.nanoTime()
        list.removeAt(list.size / 2)
        return System.nanoTime() - startTime
    }

    private fun removeEnd(list: MutableList<String>): Long {
        val startTime: Long = System.nanoTime()
        list.removeAt(list.size - 1)
        return System.nanoTime() - startTime
    }
}
